### PR TITLE
Allow the jaeger exporter path to be configured

### LINKED ISSUE #2821: Allow us to specify the path to post telemetry to in the JaegerHttpClient
# Feature Request

**Is your feature request related to a problem?**

In the `JaegerHttpClient`, the path to post is hard-coded to `/api/traces`: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerHttpClient.cs#L70
I'd like to be able to specify this because the way our infrastructure is setup is that the listener is expecting a different path. And it won't collect trace data if that is appended. We're switching from the Jaeger library with OpenTracing, and it worked previously. I see that the zipkin exporter allows you to specify the endpoint directly

**Describe the solution you'd like:**

I'd either like to be able to do the same, or have this hard-coded path be configurable.

**Describe alternatives you've considered.**

I can use the zipkin exporter, but since we've been using Jaeger already, I'd like to stick with it (plus all the other non .net services still use Jaeger).

## Additional Context

I'm happy to implement the feature myself, but thought I should open it up for discussion before doing so.

### PR DESCRIPTION
**Pull request by abe545**
Fixes #2821.

## Changes

Please provide a brief description of the changes here.
This makes the `JaegerExporter` treat the configurable `Endpoint` in its options the same way the `ZipkinExporter` does. Basically, instead of hardcoding `/api/traces` as a path to `POST` the trace data to, it is now assumed to be a part of the `Endpoint`. This is the cleanest way of implementing this feature, but it is possible it will break people who already have it configured correctly. 

I'm open to other ways of doing this - I'm not sure what the ethos is in this project surrounding changes like this. As I mentioned in the original feature request, I have some other ideas about how to achieve the goal (the goal being to be able to fully control the URL being used to post trace data), but figured I'd start with the cleanest code and move from there.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [x] Design discussion issue # (https://github.com/open-telemetry/opentelemetry-specification/issues/2331)
* [ ] Changes in public API reviewed

### CODE REVIEW

File: test/OpenTelemetry.Exporter.Jaeger.Tests/OpenTelemetry.Exporter.Jaeger.Tests.csproj
Comment (cijothomas) [inline test/OpenTelemetry.Exporter.Jaeger.Tests/OpenTelemetry.Exporter.Jaeger.Tests.csproj:21]: we try to avoid adding extra packages, even in test projects. do we really this one?

Comment (abe545) [reply test/OpenTelemetry.Exporter.Jaeger.Tests/OpenTelemetry.Exporter.Jaeger.Tests.csproj:21]: It is the easiest way I know to test an http client - I see the zipkin exporter tests use a `TestHttpServer`. I'll switch this test to use it.

Comment (cijothomas) [reply test/OpenTelemetry.Exporter.Jaeger.Tests/OpenTelemetry.Exporter.Jaeger.Tests.csproj:21]: Thanks.

**Comment by cijothomas** [comment]
> [CLA Not Signed](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/185478059/2847/#/?version=2)
> 
> * [❌](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/185478059/2847/#/?version=2) - Abraham Heidebrecht The commit ([6e0ce44](https://github.com/open-telemetry/opentelemetry-dotnet/commit/6e0ce44443d453d0ffb5f787c0064cb99805b90d)) is not authorized under a signed CLA. [Please click here to be authorized](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/185478059/2847/#/?version=2). For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

@abe545 Also, please sign the SLA, before we can consider contributions!

**Comment by abe545** [comment]
> > [CLA Not Signed](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/185478059/2847/#/?version=2)
> > 
> > * [❌](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/185478059/2847/#/?version=2) - Abraham Heidebrecht The commit ([6e0ce44](https://github.com/open-telemetry/opentelemetry-dotnet/commit/6e0ce44443d453d0ffb5f787c0064cb99805b90d)) is not authorized under a signed CLA. [Please click here to be authorized](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/185478059/2847/#/?version=2). For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).
> 
> @abe545 Also, please sign the SLA, before we can consider contributions!

Thanks - I'm trying to figure out who the CLA manager is at my organization, and get them to sign it.

File: test/OpenTelemetry.Exporter.Jaeger.Tests/OpenTelemetry.Exporter.Jaeger.Tests.csproj
Comment (abe545) [reply test/OpenTelemetry.Exporter.Jaeger.Tests/OpenTelemetry.Exporter.Jaeger.Tests.csproj]: No problem at all

File: test/OpenTelemetry.Exporter.Jaeger.Tests/OpenTelemetry.Exporter.Jaeger.Tests.csproj
Comment (reyang) [inline test/OpenTelemetry.Exporter.Jaeger.Tests/OpenTelemetry.Exporter.Jaeger.Tests.csproj:11]: Consider moving this to line 33.

**Review by reyang — APPROVED** [review]
LGTM with minor suggestions.

File: test/OpenTelemetry.Exporter.Jaeger.Tests/OpenTelemetry.Exporter.Jaeger.Tests.csproj
Comment (abe545) [reply test/OpenTelemetry.Exporter.Jaeger.Tests/OpenTelemetry.Exporter.Jaeger.Tests.csproj:11]: Oh man, I don't know how I missed those lines that were already there 🙄

**Review by reyang — APPROVED** [review]

**Review by mic-max — APPROVED** [review]

File: src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs:87]: I think this is a breaking behavior. Previously /api/traces was appended to the endpoint user provided. Now we expect the user to provide the whole url.

(should be fine as Endpoint was added only in 1.2 and we haven;t yet released a stable) 

@CodeBlanch Could you comment?

Also, we should update the readme to reflect the new change.

**Review by cijothomas — DISMISSED** [review]
Might be breaking, so want to make sure we are okay with this.
https://github.com/open-telemetry/opentelemetry-dotnet/pull/2847/files#r799001655

File: src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs
Comment (abe545) [reply src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs:87]: Agreed that this a breaking change. As I said in my PR comment, I personally think it makes sense to have this work like the zipkin exporter does - but I totally understand if you would like to avoid a breaking change. This is fairly trivial to change to some other pattern (like maybe another property on the `JaegerExporterOptions` with the path to append to the endpoint, and have it default to `/api/traces`).
I'll update the readme as a part of this PR to reflect the decision you make on how to implement this.

Comment (abe545) [reply src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs:87]: Hi @cijothomas just wanted to check if you've had a chance to discuss with the other maintainers what you'd like me to do here? Thanks!

Comment (pellared) [reply src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs:87]: It may be good to confront it with the specification and how other languages are doing it (I suspect inconsistency).

Comment (abe545) [reply src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs:87]: So in the [Environment variable docs](https://opentelemetry.io/docs/reference/specification/sdk-environment-variables/#jaeger-exporter) it just lists `http://localhost:14250` as the default value, and the [`jaeger` docs](https://www.jaegertracing.io/docs/1.31/apis/#thrift-over-http-stable) are explicit about the path to use for Thrift over HTTP: `/api/traces`. I guess where this breaks down for my organization is that we don't use the jaeger agent, but rather a third party one that fails when we use this combination of values.

I've done some investigation on other otel languages:
* Java - it looks like it only supports grpc
* Javascript - Namely is already successfully sending metrics to our agent using (It looks like otel js is using the `jaeger-client` npm package, which just uses the `endpoint` as-is):
```js
const exporter = new JaegerExporter({
  endpoint: 'http://our-trace-collector:9080/v1/trace',
});
```
* Golang - according to their [readme](https://github.com/open-telemetry/opentelemetry-go/tree/v1.3.0/exporters/jaeger#configuration) and [docs](https://pkg.go.dev/go.opentelemetry.io/otel/exporters/jaeger#WithEndpoint), the endpoint can be specified with a path (`OTEL_EXPORTER_JAEGER_ENDPOINT` defaults to `http://localhost:14268/api/traces`), and looking through the code, it just uses the `endpoint` value specified
* Ruby - [readme](https://github.com/open-telemetry/opentelemetry-go/blob/bd817df68c274fbace901fab63da2bea267315ab/exporters/jaeger/uploader.go#L294) and [code](https://github.com/open-telemetry/opentelemetry-ruby/blob/980f0e454c59e822de439635ee513e2bcdf820b8/exporter/jaeger/lib/opentelemetry/exporter/jaeger/collector_exporter.rb#L28) indicate that they export to the URL specified in the env var as-is

I haven't looked through any of the other languages, but it does seem that my change is more consistent with at least a few of the other languages that otel supports (and incidentally all the ones my Company uses 😄).

Comment (pellared) [reply src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs:87]: @abe545 I think that .NET SDK does it correctly. You may try discussing it with the [OTel spec SIG](https://github.com/open-telemetry/community#special-interest-groups) if it bothers you. Especially that Jaeger exporter is a quite popular one and having inconsistencies between languages can be frustrating.

Comment (pellared) [reply src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs:87]: FYI https://github.com/open-telemetry/opentelemetry-go/issues/2599

Comment (abe545) [reply src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs:87]: One other thing I forgot to mention is that the behavior in this repo is different from the [OpenTracing jaeger client](https://github.com/jaegertracing/jaeger-client-csharp) (which we're migrating from). It also just used the endpoint property/env-var as-is.
If you want to keep the behavior as-is, would you have a problem with me adding a new configuration option to set the path? Right now there is no way for me to use jaeger with open telemetry in my organization for .net. That is the problem I'm trying to solve, and I'm not terribly concerned about *how* to solve it.

Comment (abe545) [reply src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs:87]: Now that it looks like the spec will be updated to make it clear that the endpoint should include the path, how do you feel about this pr @pellared and @cijothomas (also, do I need to do something to kick off the build)?

Comment (pellared) [reply src/OpenTelemetry.Exporter.Jaeger/JaegerExporterOptions.cs:87]: @abe545 In general I prefer the approach of your PR (this is also how I implemented it in OTel Go SDK some time ago 😄). The change is not really breaking from the perspective as there was no stable release.

**Comment by cijothomas** [comment]
@abe545 The change looks good to me. Please address the following before merge:
1. Please update changelog to include this. (highlight that this will break previous rc users, so they need to take action).
2. Update PR Description to include the relevant Spec links.

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Comment by abe545** [comment]
> @abe545 The change looks good to me. Please address the following before merge:
> 
> 1. Please update changelog to include this. (highlight that this will break previous rc users, so they need to take action).
> 2. Update PR Description to include the relevant Spec links.

Thanks @cijothomas! I updated the changelog and PR comment - please let me know if you'd like me to make any other changes

File: src/OpenTelemetry.Exporter.Jaeger/CHANGELOG.md
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Jaeger/CHANGELOG.md:9]: Please add a link to this pr or the issue as well.

Comment (abe545) [reply src/OpenTelemetry.Exporter.Jaeger/CHANGELOG.md]: Thanks. I also linked to the PR that updated the spec.

**Comment by cijothomas** [comment]
@abe545 Could you merge the latest main to this as well?

**Review by cijothomas — APPROVED** [review]

**Pull request merged**
